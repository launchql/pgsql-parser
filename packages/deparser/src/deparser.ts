import { Node } from '@pgsql/types';
import { SqlFormatter } from './utils/sql-formatter';
import { ConsolidatedVisitor } from './visitors/consolidated-visitor';
import { DeparserContext } from './visitors/base';

export interface DeparserOptions {
  newline?: string;
  tab?: string;
}

export class Deparser {
  private formatter: SqlFormatter;
  private consolidatedVisitor: ConsolidatedVisitor;
  private tree: Node[];

  constructor(tree: Node | Node[], opts: DeparserOptions = {}) {
    this.formatter = new SqlFormatter(opts.newline, opts.tab);
    this.consolidatedVisitor = new ConsolidatedVisitor(this.formatter, this);
    
    this.tree = Array.isArray(tree) ? tree : [tree];
  }

  static deparse(query: Node | Node[], opts: DeparserOptions = {}): string {
    return new Deparser(query, opts).deparseQuery();
  }

  deparseQuery(): string {
    return this.tree
      .map(node => this.deparse(node))
      .join(this.formatter.newline() + this.formatter.newline());
  }

  deparse(node: Node, context: DeparserContext = {}): string | null {
    if (node == null) {
      return null;
    }

    if (typeof node === 'number' || node instanceof Number) {
      return node.toString();
    }

    try {
      return this.consolidatedVisitor.visit(node, context);
    } catch (error) {
      const nodeType = Object.keys(node)[0];
      throw new Error(`Error deparsing ${nodeType}: ${(error as Error).message}`);
    }
  }


}
