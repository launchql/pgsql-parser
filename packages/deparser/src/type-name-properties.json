{
  "TypeCast": ["typeName"],
  "RangeTableFuncCol": ["typeName"],
  "ColumnDef": ["typeName"],
  "XmlSerialize": ["typeName"],
  "JsonOutput": ["typeName"],
  "JsonTableColumn": ["typeName"],
  "CreateDomainStmt": ["typeName"],
  "CreateTransformStmt": ["type_name"],
  "CreateStmt": ["ofTypename"],
  "CreateOpClassItem": ["datatype", "storedtype"],
  "CreateFunctionStmt": ["returnType"],
  "FunctionParameter": ["argType"],
  "CreateCastStmt": ["sourcetype", "targettype"]
} 