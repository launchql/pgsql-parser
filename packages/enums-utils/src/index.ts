import toInt from './enums2int.json';
import toStr from './enums2str.json';
import nodes from './nodes.json';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
