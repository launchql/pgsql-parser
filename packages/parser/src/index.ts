import { Deparser, deparse } from 'pgsql-deparser';
import {
  parseQuery,
  parseQuerySync,
  parsePlPgSQLSync as parseFunction
} from 'libpg-query';

function mapStmt({ stmt, stmt_len, stmt_location }) {
  return {
    RawStmt: {
      stmt,
      stmt_len,
      stmt_location: stmt_location || 0
    }
  };
}

export const parse = (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = parseQuerySync(sql);
  return result.stmts.map(mapStmt);
};

export const parseAsync = async (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = await parseQuery(sql);
  return result.stmts.map(mapStmt);
};

export { deparse, Deparser, parseFunction };
