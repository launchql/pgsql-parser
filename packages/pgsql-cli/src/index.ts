#!/usr/bin/env node
import minimist from 'minimist';
import chalk from 'chalk';
import { readFileSync } from 'fs';
import { resolve } from 'path';

import { parseCommand } from './commands/parse';
import { deparseCommand } from './commands/deparse';
import { protoGenCommand } from './commands/proto-gen';
import { protoFetchCommand } from './commands/proto-fetch';
import { runtimeSchemaCommand } from './commands/runtime-schema';
import { showHelp, showVersion } from './utils/help';

const argv = minimist(process.argv.slice(2), {
  alias: {
    h: 'help',
    v: 'version',
    o: 'output',
    f: 'format'
  }
});

const command = argv._[0];

async function main() {
  try {
    if (argv.version) {
      showVersion();
      process.exit(0);
    }

    if (!command || argv.help) {
      showHelp(command);
      process.exit(0);
    }

    switch (command) {
      case 'parse':
        await parseCommand(argv);
        break;
      
      case 'deparse':
        await deparseCommand(argv);
        break;
      
      case 'proto-gen':
        await protoGenCommand(argv);
        break;
      
      case 'proto-fetch':
        await protoFetchCommand(argv);
        break;
      
      case 'runtime-schema':
        await runtimeSchemaCommand(argv);
        break;
      
      default:
        console.error(chalk.red(`Unknown command: ${command}`));
        showHelp();
        process.exit(1);
    }
  } catch (error: any) {
    console.error(chalk.red('Error:'), error.message || error);
    process.exit(1);
  }
}

main();
