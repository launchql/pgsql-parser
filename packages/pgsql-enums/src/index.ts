import toInt from './enums2int';
import toStr from './enums2str';
import nodes from './nodes';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
