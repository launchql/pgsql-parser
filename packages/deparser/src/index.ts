import { Deparser } from "./deparser";

const deparse = Deparser.deparse;
export { deparse, Deparser };
