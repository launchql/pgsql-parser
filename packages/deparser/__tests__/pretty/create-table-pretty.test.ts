import { deparseSync } from '../../src';
import { parse } from 'libpg-query';

describe('Pretty CREATE TABLE formatting', () => {
  const basicTableSql = `CREATE TABLE users (id SERIAL PRIMARY KEY, name TEXT NOT NULL, email TEXT UNIQUE);`;
  
  const complexTableSql = `CREATE TABLE orders (
    id SERIAL PRIMARY KEY,
    user_id INTEGER NOT NULL,
    total DECIMAL(10,2) CHECK (total > 0),
    status VARCHAR(20) DEFAULT 'pending',
    created_at TIMESTAMP DEFAULT now(),
    FOREIGN KEY (user_id) REFERENCES users(id)
  );`;

  it('should format basic CREATE TABLE with pretty option enabled', async () => {
    const parsed = await parse(basicTableSql);
    const result = deparseSync(parsed, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should maintain single-line format when pretty option disabled', async () => {
    const parsed = await parse(basicTableSql);
    const result = deparseSync(parsed, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should format complex CREATE TABLE with pretty option enabled', async () => {
    const parsed = await parse(complexTableSql);
    const result = deparseSync(parsed, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should maintain single-line format for complex table when pretty disabled', async () => {
    const parsed = await parse(complexTableSql);
    const result = deparseSync(parsed, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should use custom newline and tab characters in pretty mode', async () => {
    const parsed = await parse(basicTableSql);
    const result = deparseSync(parsed, { 
      pretty: true, 
      newline: '\r\n', 
      tab: '    ' 
    });
    expect(result).toMatchSnapshot();
  });
});
