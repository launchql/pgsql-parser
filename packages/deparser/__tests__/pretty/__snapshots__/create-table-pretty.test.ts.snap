// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pretty CREATE TABLE formatting should format basic CREATE TABLE with pretty option enabled 1`] = `
"CREATE TABLE users (
  id serial PRIMARY KEY,
  name text NOT NULL,
  email text UNIQUE
);"
`;

exports[`Pretty CREATE TABLE formatting should format complex CREATE TABLE with pretty option enabled 1`] = `
"CREATE TABLE orders (
  id serial PRIMARY KEY,
  user_id int NOT NULL,
  total numeric(10, 2) CHECK (total > 0),
  status varchar(20) DEFAULT 'pending',
  created_at pg_catalog.timestamp DEFAULT now(),
  FOREIGN KEY (user_id) REFERENCES users (id)
);"
`;

exports[`Pretty CREATE TABLE formatting should maintain single-line format for complex table when pretty disabled 1`] = `"CREATE TABLE orders (id serial PRIMARY KEY, user_id int NOT NULL, total numeric(10, 2) CHECK (total > 0), status varchar(20) DEFAULT 'pending', created_at pg_catalog.timestamp DEFAULT now(), FOREIGN KEY (user_id) REFERENCES users (id));"`;

exports[`Pretty CREATE TABLE formatting should maintain single-line format when pretty option disabled 1`] = `"CREATE TABLE users (id serial PRIMARY KEY, name text NOT NULL, email text UNIQUE);"`;

exports[`Pretty CREATE TABLE formatting should use custom newline and tab characters in pretty mode 1`] = `
"CREATE TABLE users (
    id serial PRIMARY KEY,
    name text NOT NULL,
    email text UNIQUE
);"
`;
